transcom/mymove
# INT B-19819 unique DODID when creating a customer

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-19819)

## Summary

When a service counselor is creating a customer, we want to make sure that the DODID that they're providing doesn't already exist. This added functionality needs to be wrapped up in a feature flag because that sounds like an inconvenience that would leave testing and developing an absolute nightmare.

You can find the merged feature flag in each environment [here.](https://github.com/transcom/milmove-feature-flags/pull/179)

This PR does the following:
- added feature flag to `envrc` and set to `false`
- added feature flag value and implemented reading value from Flipt
- updated `customer.go` handler to check for existing DODID only if the feature flag is on
- added a test and updated existing test structure
- updated UI styling to server error banner to share the same width as the form
- throwing in a fix for a flaky test that was checking an array that sometimes was out of order of the comparison

### How to test

1. Run `make client_run`
2. In the `development.features.yaml` file, change the `dodid_unique` values from `false` to `true`
3. In a separate terminal, run `make feature_flag_docker`
4. In a separate terminal, run your server to mimic Flipt functionality by running <br />`FEATURE_FLAG_SERVER_URL=http://localhost:9080 make server_run`
5. Log in as a service counselor
6. Navigate to the `Customer Search` tab
7. Click the `Customer name` radio button
8. Type in literally anything and click `Search`
9. The `Add Customer` button should pop out at ya
10. Click it
11. Fill out the form and use an existing DODID value in the `DoD ID number` text input
12. Try to submit the form
13. You should see an error at the top of the page that says the DODID is already taken

## Screenshots
![Screenshot 2024-07-24 at 2 42 10 PM](https://github.com/user-attachments/assets/bb956eb3-2222-419b-810b-f098da5cec28)
